### PR TITLE
Add the buffer interface for wrapped STL vectors

### PR DESCRIPTION
**Pull request by patstew**
Allows use of vectors as python buffers, so for example they can be adopted without a copy by numpy.asarray
Allows faster conversion of numeric buffers to vectors with memcpy instead of individually casting the elements

### CODE REVIEW

**Comment by aldanor** [comment]
FYI: note that you can already wrap vectors as numpy arrays *without* copying, kind of like so:

```cpp
struct Data {
    std::vector<int> vec;
};

py::class_<Data>(m, "Data")
    .def("vec", [](Data& self) -> py::array {  // no copying
        return {self.vec.data(), self.vec.size(), py::cast(self);
    });
```

Untested, but something like this should work.

**Comment by patstew** [comment]
That's cool, I didn't know that. But going that way would require me to write a load of little wrapping lambdas for my functions, or change my C++ API to return those proxy objects. This way seems easier to me, and the buffer interface has more general use without necessarily using numpy.

**Comment by wjakob** [comment]
Neat!

@aldanor: the advantage of this version is that the original instance exposes all the necessary interfaces. (i.e. no need to call an extra function like ``.vec()``).

The restriction to arithmetic-only types if of course the main disadvantage - @aldanor's NumPy interface can handle much fancier structured data types.

I wonder if there is a way to expose a NumPy interface (analogous to the buffer interface) that works for more complex structured types, and without calling an extra conversion function?

**Comment by patstew** [comment]
I guess it would be possible to write some template machinery to let you explicitly declare members of a struct, or use a std::tuple, and match the types to the python format. But I'd guess that you'd run into problems with C++ and python disagreeing about alignment in those cases? It might be OK for a `Point { float x, float y }` sort of thing, but one you start mixing types are there any memory layout guarantees?

**Comment by aldanor** [comment]
@patstew That's exactly what "record array" part of `<pybind11/numpy.h>` does :) E.g. figures out the exact offsets, alignment, supports nested structs, etc, but you have to call a macro and pass it the field names. A fairly good demo of supported types is [in the tests](https://github.com/pybind/pybind11/blob/master/tests/test_numpy_dtypes.cpp). 

Why the "is_arithmetic" restriction? I don't why this would work with structured dtypes. `PYBIND11_NUMPY_DTYPE` registeres both `npy_format_descriptor<>` and `format_descriptor<>`, so it should work in this case just as well?

Re: "not bool", this bit looks a tad flaky; I guess what you really want to check is that the container type is not a proxy and underlying memory is contiguous. Simply checking that `.data()` returns a pointer is sufficient for STL containers, I think (`vector<bool>` doesn't have it in the specialization).

**Comment by patstew** [comment]
There we go, it now goes to and from structured arrays too. I've stopped pybind11 from declaring aligned types as unaligned in the format string and added a function to compare format strings, so that I can tell when a numpy structured dtype and a c++ one are the same, even when the format strings are not literally the same. numpy tends to declare padding as 'xxx' rather than '3x' and doesn't add trailing padding to the format string. Also, at least for me on windows, numpy and c++ don't agree on whether int32 is int or long so I allow 'i' == 'l' when appropriate.

**Comment by aldanor** [comment]
Ok, so it looks like this is now 2 (3?) separate issues, and as such it may even make sense to break it down into separate PRs.

1) Recording alignment of fields for structured dtypes. Makes sense to have -- however it would be also nice to add a test that actually *fails* without this (i.e. at the current master branch) at least on one platform. (We also ignore byte ordering at the moment btw)

2) The actual vector/buffer interface

Re: `compare_format_descriptor()`, congrats, you've started partially reimplementing functionality from [`numpy.core._internal`](https://github.com/numpy/numpy/blob/master/numpy/core/_internal.py#L423) :) While your code is likely to be correct (not clear; needs extensive tests if this is to be merged in), tbh I wouldn't be a huge fan of maintaining it. Buffer format strings are horrible, and there's always a myriad of weird corner cases.

One alternative option would be to rely on numpy for comparing structured types (you have to register them *through* numpy anyway, so this makes sense). You'll still need special-casing even for arithmetic types, but now that would be a lot simpler. Another option is to use numpy dtypes for comparing all underlying value types; this imposes reliance on numpy but would be a bit cleaner.

There's already a `py::dtype(const buffer_info&)` ctor, and there's `py::dtype::of<T>()` for all registered types. Dtypes can then be compared for equivalence (not equality), which should handle the corner cases like `sizeof(int) == sizeof(long)` on Windows platforms (will still need tests).

**Comment by aldanor** [comment]
@wjakob @patstew

> @aldanor: the advantage of this version is that the original instance exposes all the necessary interfaces. (i.e. no need to call an extra function like .vec()).

Well, there's [NumPy array interface](https://docs.scipy.org/doc/numpy/reference/arrays.interface.html) which is very simple to implement and doesn't require dancing around Python buffer protocol and its format strings. Then the original instance would also "expose all the necessary interfaces" as you could just `np.asarray(vec)`, or even just pass it in places where numpy array is expected.

**Comment by patstew** [comment]
Ok, I've cut out some of the extra stuff for this PR. It basically means that the fast __init__ path won't work as often, but exposing vectors works fine. I've added some non-numpy tests too.
My original reason for adding this was so that passing around big `vector<char>`s (from file/network/etc) wasn't doing quite so much unnecessary work. The numpy aspect was a nice to have, and handy for testing across py2/py3. I'd prefer not to tie the whole thing to numpy.

**Comment by aldanor** [comment]
@patstew So, it looks like the cases are:

1. `T` is a non-structured type. In this case you can compare format strings, and even handle some platform-dependent weirdness manually (like `i` and `l` on Windows). Maybe it'd make sense to have a `normalize_scalar_format()` function that would handle all that instead of a comparison function? With a bit of effort, you could probably even make checks like `sizeof(int) == sizeof(long)` compile-time.

2. `T` is a structured type (basically, format string is not a single character?). Currently, you can only register those via `PYBIND11_NUMPY_DTYPE`, so we know that NumPy is present, and we can compare dtypes for equivalence, namely `py::dtype::of<T>()` and dtype constructed from buffer info object.

Note that you could still use NumPy dtype comparison in (1), but if comparing format strings for simple scalars is sufficient then it's fine.

**Comment by aldanor** [comment]
Here's another thought, maybe there could be a separate type for format strings, e.g. `buffer_format`, implicitly convertible from/to `std::string` so as not to break existing code, and implementing comparison op following the logic above. This way it could be more reusable, since it's not stl-vector-specific inherently.

**Comment by patstew** [comment]
Using the numpy comparison function seems to fail when there's trailing padding required, because numpy doesn't include it and counts it as different. I suspect it's to do with this https://github.com/numpy/numpy/pull/7798

**Comment by aldanor** [comment]
@patstew Yes, you can notice my comments in https://github.com/numpy/numpy/issues/7797 and https://github.com/numpy/numpy/issues/7798.

This is why we have these bits of code in pybind11:  
- https://github.com/pybind/pybind11/blob/master/include/pybind11/numpy.h#L771
- https://github.com/pybind/pybind11/blob/master/include/pybind11/numpy.h#L272

**Comment by aldanor** [comment]
(as in, if the padding doesn't get stripped properly by `py::dtype`, it's a bug we have to fix regardless; I'll try to take a look at that branch)

File: include/pybind11/stl_bind.h [hunk 1]
```diff
@@ -326,6 +327,36 @@ template <typename Vector, typename Class_> auto vector_if_insertion_operator(Cl
     );
 }
 
+// Provide the buffer interface for vectors if we have data() and we have a format for it
+// GCC seems to have "void std::vector<bool>::data()" - doing SFINAE on the existence of data() is insufficient, we need to check it doesn't return void
```
Comment (wjakob) [inline include/pybind11/stl_bind.h]: Not sure I understand this? Are we treating GCC specially here?

Comment (wjakob) [inline include/pybind11/stl_bind.h]: If the goal is just to detect the ``bool`` special case, wouldn't it be more transparent to SFINAE on that?

File: include/pybind11/stl_bind.h [hunk 2]
```diff
@@ -326,6 +327,36 @@ template <typename Vector, typename Class_> auto vector_if_insertion_operator(Cl
...
+template <typename Vector, typename Class_>
+enable_if_t<!std::is_same<decltype(py::format_descriptor<typename Vector::value_type>::format(), std::declval<Vector>().data()), void>::value>
+vector_buffer(Class_& cl) {
+    using T = typename Vector::value_type;
+
+    try {
+        //numpy.h declares this for arbitrary types, but it may raise an exception if PYBIND11_NUMPY_DTYPE hasn't been called
+        py::format_descriptor<T>::format();
```
Comment (patstew) [inline include/pybind11/stl_bind.h]: It works for the ordinary arithmetic types declared in `common.h`, or fails to compile because the default implementation of `py::format_descriptor` is empty. `numpy.h` provides a specialisation for all POD types, which may throw if the type isn't declared properly.

File: include/pybind11/stl_bind.h (hunk 1 above)
Comment (patstew) [inline include/pybind11/stl_bind.h]: I did that originally, aldanor said he'd prefer checking for data().

Comment (wjakob) [inline include/pybind11/stl_bind.h]: cc @aldanor

File: include/pybind11/stl_bind.h (hunk 2 above)
Comment (wjakob) [inline include/pybind11/stl_bind.h]: Ok -- this sounds good to me then.

**Comment by wjakob** [comment]
Note: this PR is currently conflicted and needs to be rebased on top of ``master``.

**Comment by patstew** [comment]
I've fixed the issue with padding in that branch (https://github.com/patstew/pybind11/tree/compare_buffer_numpy) now. I'll submit it as a new PR once this has gone in, unless you'd like me to add it to this PR.

**Comment by aldanor** [comment]
@patstew Yes, the fix/hack looks correct to me; I actually fixed it almost exactly the same way locally while looking at why your branch fails.

Would you mind opening a separate PR just for that? (+ a test in `test_numpy_dtypes` that would be broken on the current master but would get fixed with the `strip_padding` fix)

File: include/pybind11/stl_bind.h (hunk 2 above)
Comment (jagerman) [inline include/pybind11/stl_bind.h]: Instead of checking that it *isn't* void, why don't you check that it *is* a `Vector::value_type *`?  (Ultimately that seems to align better with what actually matters in the code)

File: include/pybind11/stl_bind.h (hunk 1 above)
Comment (jagerman) [inline include/pybind11/stl_bind.h]: Instead of checking that it *isn't* void, why don't you check that it *is* a `Vector::value_type *`?  (Ultimately that seems to align better with what actually matters in the code)

**Comment by patstew** [comment]
Appveyor seems to have taken an hour and not actually started building at all...

File: include/pybind11/stl_bind.h (hunk 1 above)
Comment (jagerman) [inline include/pybind11/stl_bind.h]: There's nothing, in theory, to prevent this from working with some std::vector-like custom class (just like `bind_vector`), so I think the check is worthwhile, but it should really just check for what we want instead of checking for *not* being what the current stdlibc++ implementation does.

**Comment by jagerman** [comment]
Three of the builds ran fine; the other looks like a false positive (appveyor problem).

**Comment by aldanor** [comment]
So with #505 and #520, I guess everything is in place to do comparisons of structured type formats via numpy (as in [the branch](https://github.com/patstew/pybind11/tree/compare_buffer_numpy))?

**Comment by patstew** [comment]
Ok, here's a version with structure comparisons added. There might be a couple of points for discussion/bikeshedding:
Should `compare_buffer_info` be part of the API as opposed to in `::detail`? If so is `compare_buffer_info<T>::compare(..)` a reasonable name? Should it be merged into `format_descriptor<T>`? Should it stay in detail but be used to implement `operator==` and `operator!=` for `format_descriptor`s and `buffer_info`s in the API?

**Comment by patstew** [comment]
Are there any remaining issues with this? I believe I've addressed everything raised so far.

**Comment by jagerman** [comment]
@aldanor - as far as I can tell, this code *does* copy:

```C++
struct Data { std::vector<int> vec; };

py::class_<Data>(m, "Data")
    .def("vec", [](Data& self) -> py::array {  // no copying
        return {self.vec.data(), self.vec.size(), py::cast(self);
    });
```
because the `array` constructor forces a copy when `base` is not set.  (I need to be able to do this for Eigen reference support as well).

**Comment by aldanor** [comment]
@jagerman But the base is set? (`py::cast(self)`)

**Comment by jagerman** [comment]
Oh, right, it is, I was misreading the arguments.  My issue is a little different, then: I don't *have* a base, because I'm returning an actual numpy array that maps into eigen data.  AFAICT it should be fine to do this (i.e. it seems to work with whatever I can throw at it), but `py::array` currently gives me no way to do that.

**Comment by aldanor** [comment]
So eigen data has "static" lifetime? You can probably provide a dummy Python object as a base that will never get recycled, or something like that.

**Comment by jagerman** [comment]
No, it's more of a you-'re-on-your-own as far as lifetime management, essentially the same thing you get with `rvp::reference`.

**Comment by patstew** [comment]
This was broken by the addition of `py::buffer_protocol()`, I've fixed that and a merge conflict. I'm not sure why AppVeyor is refusing to build it?

**Comment by dean0x7d** [comment]
The CI hooks may not have triggered correctly. Try pushing a new commit and see if it recovers.

**Comment by patstew** [comment]
@wjakob @aldanor @jagerman Are there any outstanding issues with this?

**Comment by jagerman** [comment]
@patstew: the PR currently disables (comments out) most of the travis-ci builds, which obviously need to be undone.

File: include/pybind11/stl_bind.h [hunk 3]
```diff
@@ -326,6 +326,48 @@ template <typename Vector, typename Class_> auto vector_if_insertion_operator(Cl
...
+    cl.def_buffer([](Vector& v) -> py::buffer_info {
+        return py::buffer_info(v.data(), sizeof(T), py::format_descriptor<T>::format(), 1, {v.size()}, {sizeof(T)});
+    });
+
+    cl.def("__init__", [](Vector& vec, py::buffer buf) {
+        auto info = buf.request();
+        if (info.ndim != 1)
+            throw pybind11::type_error("Only 1D buffers can be copied to a vector");
+        if (info.strides[0] != sizeof(T))
+            throw pybind11::type_error("Item size mismatch (Python: " + std::to_string(info.strides[0]) + " C++: " + std::to_string(sizeof(T)) + ")");
```
Comment (jagerman) [inline include/pybind11/stl_bind.h]: I don't think this `strides[0] == sizeof(T)` requirement is desirable: it should be easy enough to support arbitrary strides, which let slices (e.g. single rows/columns of numpy matrices) work as vector buffer inputs.  Something like this (untested) as a replacement for the Vector constructor ought to work:

```C++
new (&vec) Vector();
vec.reserve(info.shape[0]);
for (void *ptr = info.ptr, void *end = info.ptr + info.strides[0] * info.shape[0]; ptr != end; ptr += info.strides[0])
    vec.push_back(*static_cast<T *>(ptr));
```

(A small test to see that this really does work would be good as well).

Comment (patstew) [inline include/pybind11/stl_bind.h]: Added with test

File: tests/test_stl_binders.py [hunk 4]
```diff
@@ -26,6 +33,54 @@ def test_vector_int():
...
+
+    a = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]], dtype=np.uintc)
+    v = VectorInt(a[0, :])
+    assert len(v) == 4
+    assert v[2] == 3
+    m = np.asarray(v)
+    m[2] = 5
+    assert v[2] == 5
+
+    v = VectorInt(a[:, 1])
```
Comment (jagerman) [inline tests/test_stl_binders.py]: Nice 👍

**Comment by wjakob** [comment]
I took a brief look at the latest iteration of this PR. I realize that it has been in the works for quite a while. Right now I'm still not 100% excited about it, for these reasons:

1. It's a quite heavy addition to add the buffer interface to a]every ``std::vector`` binding. I'm concerned about codebases which use many different kinds of vectors and potentially don't even need this feature. It would be good if this is opt-in.

2. Some parts of this patch could be implemented more elegantly building on top of the existing ``array_t<>``.I'm talking about stuff like

```cpp
for (char *p = static_cast<char*>(info.ptr), *end = static_cast<char*>(info.ptr) + info.shape[0] * info.strides[0]; p < end; p += info.strides[0])
            vec.push_back(*reinterpret_cast<T*>(p));
```
In practice, I suppose that non-trivial types will require NumPy support in any case. 

Thoughts?

**Comment by patstew** [comment]
I've made it conditional on the existing py::buffer_protocol() option.
I've only just added that particular monstrosity to deal with strided arrays as @jagerman requested. I'd prefer not to make this wholly tied to numpy or array_t, my original motivation was that passing vector<char>s of several MBs for IO was taking a noticeable amount of time going through all the cast functions. The more advanced numpy structs and strides stuff is there because you guys requested it :).

**Comment by jagerman** [comment]
That monstrosity could probably be made much more readable if the `*end = static_cast<char*>(info.ptr) + info.shape[0] * info.strides[0]` was before the for loop instead of as a double initializer.  (I take the blame; that was copied directly from my suggestion).

**Comment by patstew** [comment]
I've got rid of the `reinterpret_cast` too, by assuming (and checking) that `strides[0]` is a multiple of `itemsize` (Python docs say it should be) and `itemsize == sizeof(T)` (which it should if the format matched).

**Comment by dean0x7d** [comment]
This seems like a very useful feature and it's been cooking for a while. Are there any blockers left?

**Comment by wjakob** [comment]
No, I think it looks great now! Thank you @patstew!

**Comment by wjakob** [comment]
(I've added you to the acknowledgements in README.md)
